File: leetcode/p00240/p00240.go
package p00240

// Time Complexity: O(m+n)
// Space Complexity: O(1)
func searchMatrix(matrix [][]int, target int) bool {
	col := len(matrix) - 1
	row := 0

	for col >= 0 && col <= len(matrix)-1 && row >= 0 && row <= len(matrix[0])-1 {
		if matrix[col][row] == target {
			return true
		} else if matrix[col][row] < target {
			row++
		} else {
			col--
		}
	}

	return false
}

// Time Complexity: O(mlogn)
// Space Complexity: O(1)
func searchMatrix_binarysearch(matrix [][]int, target int) bool {
	for _, arr := range matrix {
		if target < arr[0] || target > arr[len(arr)-1] {
			continue
		}

		head := 0
		tail := len(matrix[0]) - 1
		for head <= tail {
			mid := head + (tail-head)/2
			if arr[mid] == target {
				return true
			} else if arr[mid] > target {
				tail = mid - 1
			} else {
				head = mid + 1
			}
		}
	}

	return false
}

File: leetcode/p00240/p00240.py
from typing import List

class Solution:
    # Time Complexity: O(m+n)
    # Space Complexity: O(1)
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        col = len(matrix)-1
        row = 0

        while col >= 0 and col <= len(matrix)-1 and row >= 0 and row <= len(matrix[0])-1:
            if matrix[col][row] == target:
                return True
            elif matrix[col][row] > target:
                col -= 1
            else:
                row += 1

        return False

if __name__ == '__main__':
    s = Solution()

    assert s.searchMatrix([
        [1,4,7,11,15],
        [2,5,8,12,19],
        [3,6,9,16,22],
        [10,13,14,17,24],
        [18,21,23,26,30]], 5) == True

    assert s.searchMatrix([
        [1,4,7,11,15],
        [2,5,8,12,19],
        [3,6,9,16,22],
        [10,13,14,17,24],
        [18,21,23,26,30]], 20) == False